src/Features/CSharp/Portable/ConvertLinq/ConvertForEachToLinqQuery/CSharpConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,340 @@
+                                }
+                            }
+
+                            // Process the last statement separately.
+                            current = array.Last();
+                        }
+                        else
+                        {
+                            // Silly case: the block is empty. Stop processing.
+                            statementsCannotBeConverted = Enumerable.Empty<StatementSyntax>();
+                        }
+
+                        break;
+
+                    case SyntaxKind.ForEachStatement:
+                        // foreach can always be converted to a from clause.
+                        var currentForEachStatement = (ForEachStatementSyntax)current;
+                        identifiersBuilder.Add(currentForEachStatement.Identifier);
+                        convertingNodesBuilder.Add(new ExtendedSyntaxNode(currentForEachStatement, currentLeadingTokens, Enumerable.Empty<SyntaxToken>()));
+                        currentLeadingTokens = new List<SyntaxToken>();
+                        // Proceed the loop with the nested statement.
+                        current = currentForEachStatement.Statement;
+                        break;
+
+                    case SyntaxKind.IfStatement:
+                        // Prepare conversion of 'if (condition)' into where clauses.
+                        // Do not support if-else statements in the conversion.
+                        var ifStatement = (IfStatementSyntax)current;
+                        if (ifStatement.Else == null)
+                        {
+                            convertingNodesBuilder.Add(new ExtendedSyntaxNode(
+                                ifStatement, currentLeadingTokens, Enumerable.Empty<SyntaxToken>()));
+                            currentLeadingTokens = new List<SyntaxToken>();
+                            // Proceed the loop with the nested statement.
+                            current = ifStatement.Statement;
+                            break;
+                        }
+                        else
+                        {
+                            statementsCannotBeConverted = new[] { current };
+                            break;
+                        }
+
+                    case SyntaxKind.LocalDeclarationStatement:
+                        // This is a situation with "var a = something;" s the innermost statements inside the loop.
+                        var localDeclaration = (LocalDeclarationStatementSyntax)current;
+                        if (localDeclaration.Declaration.Variables.All(variable => variable.Initializer != null))
+                        {
+                            // Prepare variable declarations to be converted into separate let clauses.
+                            ProcessLocalDeclarationStatement(localDeclaration);
+                            statementsCannotBeConverted = Enumerable.Empty<StatementSyntax>();
+                        }
```

Comment:
Can possibly be extracted out into `TryProcessLocalDeclaration` so it can be shared with identical code dealing with `LocalDeclarationStatementSyntax` in first switch case.